aws/resource_aws_transfer_user_test.go

```diff
@@ -0,0 +1,370 @@
+			ServerId: aws.String(serverID),
+			UserName: aws.String(userName),
+		})
+
+		if err != nil {
+			return err
+		}
+
+		*res = *describe.User
+
+		return nil
+	}
+}
+
+func testAccAWSTransferUserImportStateIdFunc(resourceName string) resource.ImportStateIdFunc {
+	return func(s *terraform.State) (string, error) {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return "", fmt.Errorf("Not found: %s", resourceName)
+		}
+
+		return fmt.Sprintf("%s/%s", rs.Primary.Attributes["user_name"], rs.Primary.Attributes["server_id"]), nil
+	}
+}
+
+func testAccCheckAWSTransferUserDisappears(serverConf *transfer.DescribedServer, userConf *transfer.DescribedUser) resource.TestCheckFunc {
+
+	return func(s *terraform.State) error {
+		conn := testAccProvider.Meta().(*AWSClient).transferconn
+
+		params := &transfer.DeleteUserInput{
+			ServerId: serverConf.ServerId,
+			UserName: userConf.UserName,
+		}
+
+		_, err := conn.DeleteUser(params)
+		if err != nil {
+			return err
+		}
+
+		return waitForTransferUserDeletion(conn, *serverConf.ServerId, *userConf.UserName)
+	}
+}
+
+func testAccCheckAWSTransferUserDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).transferconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_transfer_user" {
+			continue
+		}
+
+		userName := rs.Primary.Attributes["user_name"]
+		serverID := rs.Primary.Attributes["server_id"]
+
+		_, err := conn.DescribeUser(&transfer.DescribeUserInput{
+			UserName: aws.String(userName),
+			ServerId: aws.String(serverID),
+		})
+
+		if isAWSErr(err, transfer.ErrCodeResourceNotFoundException, "") {
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func testAccAWSTransferUserConfig_basic(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_transfer_server" "foo" {
+	identity_provider_type = "SERVICE_MANAGED"
+	
+	tags {
+		NAME     = "tf-acc-test-transfer-server"
+	}
+}
+
+
+resource "aws_iam_role" "foo" {
+	name = "tf-test-transfer-user-iam-role-%s"
+  
+	assume_role_policy = <<EOF
+{
+	"Version": "2012-10-17",
+	"Statement": [
+		{
+		"Effect": "Allow",
+		"Principal": {
+			"Service": "transfer.amazonaws.com"
+		},
+		"Action": "sts:AssumeRole"
+		}
+	]
+}
+EOF
+}
+
+resource "aws_iam_role_policy" "foo" {
+	name = "tf-test-transfer-user-iam-policy-%s"
+	role = "${aws_iam_role.foo.id}"
+	policy = <<POLICY
+{
+	"Version": "2012-10-17",
+	"Statement": [
+		{
+			"Sid": "AllowFullAccesstoS3",
+			"Effect": "Allow",
+			"Action": [
+				"s3:*"
+			],
+			"Resource": "*"
+		}
+	]
+}
+POLICY
+}
+
+
+resource "aws_transfer_user" "foo" {
+	server_id      = "${aws_transfer_server.foo.id}"
+	user_name      = "tftestuser"
+	role           = "${aws_iam_role.foo.arn}"
+
+	tags {
+		NAME = "tftestuser"
+	}
+}
+
+
+	`, rName, rName)
+}
+
+func testAccAWSTransferUserConfig_update(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_transfer_server" "foo" {
```

Comment:
Nit: I would recommend moving the common server and IAM role related resources to a "base" configuration function that can be included in the other test configuration functions to simplify the setup.

Reply:
I split test for option params.
But IAM Role is Requirement param, so left on base configuration.
https://docs.aws.amazon.com/transfer/latest/userguide/API_CreateUser.html#API_CreateUser_RequestSyntax